backend/src/main/java/com/redhat/cloud/notifications/db/repositories/BehaviorGroupRepository.java

```diff
@@ -357,20 +359,34 @@ public Status updateBehaviorGroupActions(String accountId, UUID behaviorGroupId,
         try {
             query.getSingleResult();
         } catch (NoResultException e) {
-            return NOT_FOUND;
+            throw new NotFoundException("Behavior group not found");
         }
 
-        // All behavior group actions that should no longer exist must be deleted.
-        String deleteQuery = "DELETE FROM BehaviorGroupAction WHERE behaviorGroup.id = :behaviorGroupId";
-        if (!endpointIds.isEmpty()) {
-            deleteQuery += " AND endpoint.id NOT IN (:endpointIds)";
-        }
-        javax.persistence.Query q = entityManager.createQuery(deleteQuery)
-                .setParameter("behaviorGroupId", behaviorGroupId);
-        if (!endpointIds.isEmpty()) {
-            q = q.setParameter("endpointIds", endpointIds);
+        /*
+         * Before changing the DB data, we need to lock the existing rows "for update" to prevent them
+         * from being modified or deleted by other transactions. Otherwise, DB deadlocks could happen.
+         */
+        String lockQuery = "SELECT endpoint.id FROM BehaviorGroupAction " +
+                "WHERE behaviorGroup.accountId = :accountId AND behaviorGroup.id = :behaviorGroupId";
+        List<UUID> actionsFromDb = entityManager.createQuery(lockQuery, UUID.class)
+                .setParameter("accountId", accountId)
+                .setParameter("behaviorGroupId", behaviorGroupId)
+                .setLockMode(PESSIMISTIC_WRITE)
+                .getResultList();
+
+        /*
+         * All behavior group actions that should no longer exist must be deleted.
+         */
+        List<UUID> actionsToDelete = new ArrayList<>(actionsFromDb);
+        actionsToDelete.removeAll(endpointIds);
+        if (!actionsToDelete.isEmpty()) {
+            String deleteQuery = "DELETE FROM BehaviorGroupAction " +
+                    "WHERE behaviorGroup.id = :behaviorGroupId AND endpoint.id IN (:actionsToDelete)";
+            entityManager.createQuery(deleteQuery)
+                    .setParameter("behaviorGroupId", behaviorGroupId)
+                    .setParameter("actionsToDelete", actionsToDelete)
+                    .executeUpdate();
         }
-        q.executeUpdate();
 
         for (UUID endpointId : endpointIds) {
```

Comment:
Do you mean this?

https://github.com/RedHatInsights/notifications-backend/blob/282570a173eb66a544c58ee02ef29fe980cd9b97/backend/src/main/java/com/redhat/cloud/notifications/db/repositories/BehaviorGroupRepository.java#L376-L381

It's already in the old code (a few lines under your comment) and not changed by this PR.

Reply:
Ah yes, that should do. I think it was probably because of reviewing the other flow that I felt the need to remove already-in-the-db entries from the loop.